django/template/defaultfilters.py

```diff
@@ -182,8 +182,11 @@ def floatformat(text, arg=-1):
     # Set the precision high enough to avoid an exception (#15789).
     tupl = d.as_tuple()
     units = len(tupl[1])
-    units += -tupl[2] if m else tupl[2]
+    units += tupl[2]
     prec = abs(p) + units + 1
+    prec = max(1, prec)
+    if d == 0:
+        prec = p
```

Comment:
Do we need all these changes? The following was enough for me to fix the added assertions:
```diff
diff --git a/django/template/defaultfilters.py b/django/template/defaultfilters.py
index 78881987fc..6009c94b3d 100644
--- a/django/template/defaultfilters.py
+++ b/django/template/defaultfilters.py
@@ -183,7 +183,7 @@ def floatformat(text, arg=-1):
     tupl = d.as_tuple()
     units = len(tupl[1])
     units += -tupl[2] if m else tupl[2]
-    prec = abs(p) + units + 1
+    prec = max(1, abs(p) + units + 1)
 
     # Avoid conversion to scientific notation by accessing `sign`, `digits`,
     # and `exponent` from Decimal.as_tuple() directly.
```
Maybe we need additional tests?

Reply:
Hi @felixxm 

Yes, for the patch I submitted it might seem a bit weird. But, I had isolated the code and noticed that the calculated precision values seemed off for various test-cases.

```python
def get_precision_django(d: Decimal, p: int) -> int:
    """Current Django code"""
    try:
        m = int(d) - d
    except (ValueError, OverflowError, InvalidOperation):
        return 0
    sign, digits, number_of_decimals = d.as_tuple()
    units = len(digits)  # total number of digits
    # if there are decimals, ...
    if m:
        units += -number_of_decimals
    else:
        units += number_of_decimals
    prec = abs(p) + units + 1
    return prec


def get_precision(d: Decimal, p: int) -> int:
    """Return the number of significant digits for a decimal number.

    d: the decimal number
    p: the number of decimals requested
    """
    if d == 0:
        # If the number is zero, the precision is the number of decimals requested
        return p
    _, digits, exponent = d.as_tuple()
    units = len(digits)  # total number of digits
    before_comma = units + exponent

    # adding 1 for possible rounding up e.g. (99.99, 1) -> 100.0 needs 4 digits
    prec = abs(p) + before_comma + 1
    return max(1, prec)  # 1 is the minimum precision for Decimal.quantize()

```

Tests for the above code:
```python
# A digit will be added due to rounding
ADDS_DIGIT = 1


@pytest.mark.parametrize(
    "d, p, expected",
    [
        ("123", 0, 3 + ADDS_DIGIT),
        ("123", 1, 4 + ADDS_DIGIT),
        ("123.1", 0, 3 + ADDS_DIGIT),
        ("123.1", 1, 4 + ADDS_DIGIT),
        ("1.23456789", 2, 3 + ADDS_DIGIT),
        ("123456.123456789012345678901", 2, 8 + ADDS_DIGIT),
        ("123456.123456789012345678901", 21, 27 + ADDS_DIGIT),
        ("0.0", 1, 1),
        ("0.00", 2, 2),
        ("0.0000", 2, 2),
        ("0.01", 2, 1 + ADDS_DIGIT),
        ("0.499", 0, 1),
        ("1.32e1", 2, 4 + ADDS_DIGIT),
        ("1.323297138040798e35", 2, 38 + ADDS_DIGIT),
        ("-1.323297138040798e35", 2, 38 + ADDS_DIGIT),
        ("1.5e-15", 20, 6 + ADDS_DIGIT),
        ("99999.999", 1, 7),
    ],
)
def test_get_precision(d, p, expected):
    assert get_precision(Decimal(d), p) == expected

```

Reply:
For example, the current code would set the precision too high in case
we were using floatformat("123.1", 0). It would need only prec==3, but the current code uses 6.

So, I tried understanding how many digits one would really need, and calculate that number of digits.

In the back of my head though, I think it would be better to allow setting the precision of rounding through configuration in the Django settings.py file instead of determining it dynamically based on the decimal and requested parameters.

But, as the current Django 4.2b1 broke my test-suite I tried fixing the problem without modifying the behavior too much.

Reply:
This patch is intended for backporting so we should limit it only to the reported regression. If there are other issues that exist before 08c5a787262c1ae57f6517d4574b54a5fcaad124 we can create a separate ticket for them.

Reply:
> The current tests in Django do not check if the number of decimals used is excessive, it only checks that it works.
> 
> Testing the calculated precision, would require factoring out that code (as I did in my standalone code).

I understand that the calculated precision may not be accurate, but if it doesn't produce an invalid output, we'd treat any changes in this area as cleanups, not bugfixes.

Reply:
As another example of the current code misbehaving is:
```
"123456.123456789012345678901"|floatformat:2
```
In this case, the actual required precision is 8, while the current code uses 51.

Reply:
> In this case, the actual required precision is 8, while the current code uses 51.

Yes, but the actual result is correct, so https://github.com/django/django/pull/16588#discussion_r1114260650.

Reply:
Using ` prec = max(1, abs(p) + units + 1)` to make sure the value is at least within the valid range for the Context
is indeed the smallest fix for the current regression.

Reply:
@takis I'd be more than happy to review the follow up changes that will improve the current precision calculations. Please open a separate ticket for this. Thanks :gift:

Reply:
> Using ` prec = max(1, abs(p) + units + 1)` to make sure the value is at least within the valid range for the Context is indeed the smallest fix for the current regression.

Are you happy to update this patch?

Reply:
Hi @felixxm 

Yes, I had just updated the pull request as requested.